itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestStatsReplicationScenarios.java

```diff
@@ -301,12 +338,106 @@ private String dumpLoadVerify(List<String> tableNames, String lastReplicationId,
     return dumpTuple.lastReplicationId;
   }
 
+  /**
+   * Run a bootstrap that will fail.
+   * @param tuple the location of bootstrap dump
+   */
+  private void failBootstrapLoad(WarehouseInstance.Tuple tuple, int failAfterNumTables) throws Throwable {
+    // fail setting ckpt directory property for the second table so that we test the case when
+    // bootstrap load fails after some but not all tables are loaded.
+    BehaviourInjection<CallerArguments, Boolean> callerVerifier
+            = new BehaviourInjection<CallerArguments, Boolean>() {
+      int cntTables = 0;
+      @Nullable
+      @Override
+      public Boolean apply(@Nullable CallerArguments args) {
+        cntTables++;
+        if (args.dbName.equalsIgnoreCase(replicatedDbName) && cntTables > failAfterNumTables) {
+          injectionPathCalled = true;
+          LOG.warn("Verifier - DB : " + args.dbName + " TABLE : " + args.tblName);
+          return false;
+        }
+        return true;
+      }
+    };
+
+    InjectableBehaviourObjectStore.setAlterTableModifier(callerVerifier);
+    try {
+      replica.loadFailure(replicatedDbName, tuple.dumpLocation);
+      callerVerifier.assertInjectionsPerformed(true, false);
+    } finally {
+      InjectableBehaviourObjectStore.resetAlterTableModifier();
+    }
+  }
+
+  private void failIncrementalLoad(WarehouseInstance.Tuple dumpTuple, int failAfterNumEvents) throws Throwable {
+    // fail add notification when updating table stats after given number of such events. Thus we
+    // test successful application as well as failed application of this event.
+    BehaviourInjection<NotificationEvent, Boolean> callerVerifier
+            = new BehaviourInjection<NotificationEvent, Boolean>() {
+      int cntEvents = 0;
+      @Override
+      public Boolean apply(NotificationEvent entry) {
+        cntEvents++;
```

Comment:
The add notification count in target may not match the number of events from source. So, better to count the number of AlterTable which changes last_repl_id parameters. It will be set once per event.

Reply:
This code has changed while working on another related comment. Again we don't need to count exact number of events. We need at least one successful event and other one unsuccessful event.

Reply:
OK